[Flang] [OpenMP] Add semantic checks for detach clause in task

Thirumalai-Shaktivel opened the pull request:
Fixes:
- Add semantic checks along with the tests
- Move the detach clause to allowedOnceClauses list in Task construct

Restrictions:\
OpenMP 5.0: Task construct
- At most one detach clause can appear on the directive.
- If a detach clause appears on the directive, then a mergeable clause cannot appear on the same directive.

OpenMP 5.2: Detach contruct
- If a detach clause appears on a directive, then the encountering task must not be a final task.
- A variable that appears in a detach clause cannot appear as a list item on a data-environment attribute clause on the same construct.
- A variable that is part of another variable (as an array element or a structure element) cannot appear in a detach clause.
- event-handle must not have the POINTER attribute.

Towards: https://github.com/llvm/llvm-project/pull/119172

Thirumalai-Shaktivel commented:
There seems to be some problem with `Check code formatting`.

I get the following locally
```console
$ git-clang-format
clang-format did not modify any files
```

The failed line seems to be within the range (80 characters):
```
              if (const parser::Name *name{parser::Unwrap<parser::Name>(obj)}) {
```

Thirumalai-Shaktivel committed:
[Flang] [OpenMP] Add semantic checks for detach clause in Task

Fixes:
- Add semantic checks along with the tests
- Move the detach clause to allowedOnceClauses list in
  Task construct

Restrictions:\
OpenMP 5.0: Task construct
- At most one detach clause can appear on the directive.
- If a detach clause appears on the directive, then a mergeable
  clause cannot appear on the same directive.

OpenMP 5.2: Detach contruct
- If a detach clause appears on a directive, then the encountering
  task must not be a final task.
- A variable that appears in a detach clause cannot appear as a
  list item on a data-environment attribute clause on the same
  construct.
- A variable that is part of another variable (as an array element or a
  structure element) cannot appear in a detach clause.
- event-handle must not have the POINTER attribute.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3711,6 +3768,30 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Linear &x) {
 }
 
+void OmpStructureChecker::Enter(const parser::OmpClause::Detach &x) {
+  // OpenMP 5.0: Task construct restrictions
+  CheckAllowedClause(llvm::omp::Clause::OMPC_detach);
+
+  // OpenMP 5.2: Detach clause restrictions
```
Same here: please do version-specific checks based on the actual version.

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3711,6 +3768,30 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Linear &x) {
+    }
+    auto type{name->symbol->GetType()};
+    if (!name->symbol->GetType()->IsNumeric(TypeCategory::Integer) ||
+        evaluate::ToInt64(type->numericTypeSpec().kind()) != 8) {
```
Yes, I felt the same. But, the only info I had was the `kind` => `8` to throw an error for the following:
```fortran
    !ERROR: The event-handle: `e` must be of type integer(kind=omp_event_handle_kind)
    !$omp task detach(e)
        x = x + 1
    !$omp end task
```

Do you know any other way to add a check for this?

kparzysz replied:
We don't have a way of checking this, unfortunately.  In other cases we limit it to checking that it has an INTEGER type, but without checking the kind.

Thirumalai-Shaktivel replied:
Yea, sure. I will add only checking the integer type.

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2739,6 +2739,59 @@ void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
   }
 
+  if (GetContext().directive == llvm::omp::Directive::OMPD_task) {
+    if (auto *d_clause{FindClause(llvm::omp::Clause::OMPC_detach)}) {
```
Good catch, I forgot to update this.

Thirumalai-Shaktivel commented:
Sure, thanks for the review. I will update and report back soon.

Thirumalai-Shaktivel committed:
Do not check for interger kind

Thirumalai-Shaktivel committed:
Fix snake_case

Thirumalai-Shaktivel committed:
Compare symbols instead of name

Thirumalai-Shaktivel committed:
Add OpenMP version based checks

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/detach02.f90:
```diff
@@ -0,0 +1,22 @@
+    integer(omp_event_handle_kind)          :: event_01, event_02
+
+    !TODO: Throw following error for the versions 5.0 and 5.1
+    !ERR: At most one DETACH clause can appear on the TASK directive
```
This check is required for both OpenMP 5.0 and 5.1

So, for that, I can only add a check for this in OMP.td (by moving the clause to `allowedOnceClauses`) 
 and call `CheckAllowedClause`. but, this leads to throwing an error for OpenMP version >= 52. If I enclose `CheckAllowedClause` within an if block (checking the version to be 50 and 51). The clause context wouldn't be set as `SetContextClauseInfo` is called by this function, this caused a problem for version >= 52.
 
 @kparzysz, do you have any suggestions here?

Thirumalai-Shaktivel replied:
This is not a blocker and can be fixed later on. Currently, we throw this error for every version with the latest commit

tblah replied:
Couldn't you call `SetContextClauseInfo` manually for versions >= 52?

Thirumalai-Shaktivel replied:
I thought I shouldn't be doing that (calling `SetContextClauseInfo`), but instead use only the `CheckAllowed`.

I will check it again and see what I can do there. Thank you for the suggestion.

Thirumalai-Shaktivel committed:
Fix formatting

Thirumalai-Shaktivel commented:
Done, @kparzysz, I addressed all of your comments

Thirumalai-Shaktivel commented:
Ping for review.

Thirumalai-Shaktivel commented:
@kparzysz, could you please review the changes again?

Thirumalai-Shaktivel committed:
Allow detach clause only once

Thirumalai-Shaktivel committed:
Add a test for openmp 52

Thirumalai-Shaktivel committed:
Merge branch 'main' into llvm/detach_01

tblah commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2961,6 +2961,63 @@ void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
+              checkVarAppearsInDataEnvClause(
+                  std::get<parser::OmpObjectList>(irClause.v.t),
+                  "IN_REDUCTION");
+            }
```
what about SHARED?

Thirumalai-Shaktivel replied:
Good catch, I don't know how I missed this.

tblah commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3627,40 +3683,45 @@ void OmpStructureChecker::CheckIsVarPartOfAnotherVar(
+  }
+}
+
+void OmpStructureChecker::CheckIsVarPartOfAnotherVar(
```
Suggestion, feel free to ignore. I think this would be clearer
```suggestion
void OmpStructureChecker::CheckVarIsNotPartOfAnotherVar(
```

Thirumalai-Shaktivel replied:
Done

tblah reviewed (commented):
I'm sorry this PR got lost. Please feel free to ping me if something does not get reviewed

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2961,6 +2961,63 @@ void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
+        const auto &detach{
+            std::get<parser::OmpClause::Detach>(detachClause->u)};
+        if (const auto *name{parser::Unwrap<parser::Name>(detach.v.v)}) {
+          if (name->symbol) {
```
All parser::Name's should have non-null symbols by now (it's an internal compiler error if they don't).

Thirumalai-Shaktivel replied:
Done

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3627,40 +3683,45 @@ void OmpStructureChecker::CheckIsVarPartOfAnotherVar(
+                    std::get_if<parser::DataRef>(&designator.u)}) {
+              if (IsDataRefTypeParamInquiry(dataRef)) {
+                context_.Say(source,
+                    "A type parameter inquiry cannot appear on the %s "
+                    "directive"_err_en_US,
```
Please make message strings be in a single line (here and in other places you've changed).  This is to make it easier to find them in the source based on the compiler output.

Thirumalai-Shaktivel replied:
Done

kparzysz reviewed (approved):
Sorry for the delay in reviewing.  Looks ok to me, but please wait for @tblah's input as well.

Thirumalai-Shaktivel committed:
Merge branch 'main' into llvm/detach_01

Thirumalai-Shaktivel committed:
Rename a function name

Thirumalai-Shaktivel committed:
Remove the check for symbol's nullptr

Thirumalai-Shaktivel committed:
CheckAllowedClause for OpenMP 50 and 51

Thirumalai-Shaktivel committed:
Fix the error message string to be in a single line

Thirumalai-Shaktivel committed:
Add check for shared clause as well

Thirumalai-Shaktivel committed:
Remove a test

Thirumalai-Shaktivel commented:
Thanks for the reviews, @kparzysz and @tblah.

@tblah, can you please do a final review on this PR? I have addressed your comments. If I missed anything, please let me know.

Thirumalai-Shaktivel committed:
Add some more tests

tblah reviewed (approved):
Thanks for the update. No need for another round of review for my nit comment.

NimishMishra reviewed (approved):
LGTM. The code formatting needs to be fixed though

Thirumalai-Shaktivel committed:
Fix clang-format

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4137,6 +4195,33 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Linear &x) {
 
+void OmpStructureChecker::Enter(const parser::OmpClause::Detach &x) {
+  unsigned version{context_.langOptions().OpenMPVersion};
+  if (version >= 52) {
+    SetContextClauseInfo(llvm::omp::Clause::OMPC_detach);
+  } else {
+    // OpenMP 5.0: 2.10.1 Task construct restrictions
+    CheckAllowedClause(llvm::omp::Clause::OMPC_detach);
+  }
```
> nit: won't CheckAllowedClause understand that this is allowed in newer standard versions?

Nope, are you suggesting to handle this in CheckAllowedClause itself?

```
At most one detach clause can appear on the directive.
```
The above restriction is available only for OpenMP 50 & 51.

I wanted this check not to be applied to versions equal to or above 52. But `CheckAllowedClause` will throw an error for all versions. I tried some methods in OMP.td, but it didn't help much. So, I took this route.

Thirumalai-Shaktivel commented:
Thanks for the reviews!
